### DISCUSSION TITLE
expired signal URL

### DISCUSSION THREAD

**Discussion by james987654321** [elsa-workflows/elsa-core#1304] [Q&A] [answered]
I am using a signal URL in a workflow to capture an approval from the end users but have a timeout with auto rejection.  If the signal URL is clicked after the timeout then it is no longer valid (the workflow has finished) and the user just sees an almost blank page - confusing to our end users.  

Is there a way to capture these "expired" signal URL and either put out a "this URL expired" message or redirect to another page?

**Comment by sfmskywalker** [comment, marked as answer]
I can think of three solutions to consider:

1. We publish an event from the `TriggerEndpoint` controller (which is invoked when using the signal URL generated by Elsa) - this allows your application to handle the response in case there was no workflow found that was waiting for this signal (which usually means the workflow has moved on to another state, such as when a timeout branch executed).
2. You implement an action filter for the `TriggerEndpoint` endpoint to figure out if the response was empty or not. If empty, write back a more appropriate response.
3. You implement your own controller entirely. This also means you need to generate the signal URL yourself.

I think something like option 1 is the best way, because it reduces the amount of code you would need to write yourself and is something more users will benefit from.

I will create an issue for that. In the meantime, you might consider going with option 2 or 3 so you don't have to wait for 1 to be implemented.

**Reply by sfmskywalker** [reply]
Created issue #1310

**Reply by james987654321** [reply]
thanks - not in a rush at the moment so will wait for option 1.

**Reply by sfmskywalker** [reply]
Option 1 is implemented, but not yet documented. It will be showcased as one of the final parts of the [blog series](https://sipkeschoorstra.medium.com/building-workflow-driven-net-applications-with-elsa-2-part-1-44e08a9ba94b
) I'm writing on Elsa.
